cryptoki/src/mechanism/kbkdf.rs

```diff
@@ -0,0 +1,371 @@
+    CK_SP800_108_DKM_LENGTH_SUM_OF_SEGMENTS, CK_SP800_108_FEEDBACK_KDF_PARAMS,
+    CK_SP800_108_ITERATION_VARIABLE, CK_SP800_108_KDF_PARAMS, CK_ULONG,
+};
+
+use crate::object::{Attribute, ObjectHandle};
+
+use super::MechanismType;
+
+/// Endianness of byte representation of data.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Endianness {
+    /// Little endian.
+    Little,
+    /// Big endian.
+    Big,
+}
+
+/// Defines encoding format for a counter value.
+///
+/// This structure wraps a `CK_SP800_108_COUNTER_FORMAT` structure.
+#[derive(Debug, Clone, Copy)]
+#[repr(transparent)]
+pub struct KbkdfCounterFormat(CK_SP800_108_COUNTER_FORMAT);
+
+impl KbkdfCounterFormat {
+    /// Construct encoding format for KDF's internal counter variable.
+    ///
+    /// # Arguments
+    ///
+    /// * `endianness` - The endianness of the counter's bit representation.
+    ///
+    /// * `width_in_bits` - The number of bits used to represent the counter value.
+    pub fn new(endianness: Endianness, width_in_bits: usize) -> Self {
+        Self(CK_SP800_108_COUNTER_FORMAT {
+            bLittleEndian: (endianness == Endianness::Little).into(),
+            ulWidthInBits: width_in_bits
+                .try_into()
+                .expect("bit width of KBKDF internal counter does not fit in CK_ULONG"),
+        })
+    }
+}
+
+/// Method for calculating length of DKM (derived key material).
+///
+/// Corresponds to CK_SP800_108_DKM_LENGTH_METHOD.
+#[derive(Debug, Clone, Copy)]
+pub enum KbkdfDkmLengthMethod {
+    /// Sum of length of all keys derived by given invocation of KDF.
+    SumOfKeys,
+    /// Sum of length of all segments of output produced by PRF in given invocation of KDF.
+    SumOfSegments,
+}
+
+/// Defines encoding format for DKM (derived key material).
+///
+/// This structure wraps a `CK_SP800_108_DKM_LENGTH_FORMAT` structure.
+#[derive(Debug, Clone, Copy)]
+#[repr(transparent)]
+pub struct KbkdfDkmLengthFormat(CK_SP800_108_DKM_LENGTH_FORMAT);
+
+impl KbkdfDkmLengthFormat {
+    /// Construct encoding format for length value of DKM (derived key material) from KDF.
+    ///
+    /// # Arguments
+    ///
+    /// * `dkm_length_method` - The method used to calculate the DKM length value.
+    ///
+    /// * `endianness` - The endianness of the DKM length value's bit representation.
+    ///
+    /// * `width_in_bits` - The number of bits used to represent the DKM length value.
+    pub fn new(
+        dkm_length_method: KbkdfDkmLengthMethod,
+        endianness: Endianness,
+        width_in_bits: usize,
+    ) -> Self {
+        Self(CK_SP800_108_DKM_LENGTH_FORMAT {
+            dkmLengthMethod: match dkm_length_method {
+                KbkdfDkmLengthMethod::SumOfKeys => CK_SP800_108_DKM_LENGTH_SUM_OF_KEYS,
+                KbkdfDkmLengthMethod::SumOfSegments => CK_SP800_108_DKM_LENGTH_SUM_OF_SEGMENTS,
+            },
+            bLittleEndian: (endianness == Endianness::Little).into(),
```

Comment:
I wonder if we want to have a `From<Endianness> for bool` conversion so that all thess could be replaced with `endianness.into()` :thinking:

Reply:
That feels a little wrong to me because there's no obvious equivalence between `Little` and `true` or `Big` and `false` 🤔 If some other type were to expect `true` for big endian instead (say, because it was a more obvious default), then this would break down. What do you think?

Reply:
> What do you think?

Good points! As a middle-ground maybe we could have something like this:

```rust
impl Endianness {
  /// Converts this instance into a boolean acceptable in `bLittleEndian` parameters.
  pub(crate) is_little_endian(&self) -> the param {
    *self == Endianness::Little
  }
}
```

`pub(crate)` means this would not leak and the code would be simpler (all comparisons to Little removed).

If you feel it's too much then I'm not bent into making it happen, too :sweat_smile: Hmm... the more I think about it the more unsure I am if this is even a good idea...

Maybe we can just resolve this conversation as if it never happened :face_in_clouds: